### ISSUE TITLE
signal_handler catch SIGPIPE and cause it again 

### ISSUE THREAD

**Issue by liujuncheng** [awslabs/amazon-kinesis-producer#78]
kinesis_producer cpu is very high when it's parent process is killed. strace output is below. Is signal_handler in signal_handler.cc catch signal SIGPIPE and WRITE_MESSAGE cause SIGPIPE again?

```
--- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=6927, si_uid=498} ---
write(2, "\n++++\n", 6)                 = -1 EPIPE (Broken pipe)
write(2, "[ERROR]\n", 8)                = -1 EPIPE (Broken pipe)
write(2, "---BEGIN ERROR----\n", 19)    = -1 EPIPE (Broken pipe)
write(2, "Signal: ", 8)                 = -1 EPIPE (Broken pipe)
write(2, "SIGPIPE", 7)                  = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "Description: ", 13)           = -1 EPIPE (Broken pipe)
write(2, "Broken pipe", 11)             = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "---BEGIN STACK TRACE---\n", 24) = -1 EPIPE (Broken pipe)
write(2, "4656540", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "/home/pfifer/producer-gh/amazon-"..., 88) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "104", 3)                      = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "_ZN3aws5utils9backtrace22stack_t"..., 50) = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "4655134", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "/home/pfifer/producer-gh/amazon-"..., 79) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "84", 2)                       = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "signal_handler", 14)          = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547554980095", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547554980095", 15)         = -1 EPIPE (Broken pipe)
write(2, "??:0 ??", 7)                  = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547540231373", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547540231373", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "write", 5)                    = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547539764114", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547539764114", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "_IO_new_file_write", 18)      = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547539766062", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547539766062", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "__GI__IO_file_xsputn", 20)    = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547539723292", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547539723292", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "fwrite", 6)                   = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "8343637", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "../../../../gcc-4.9.3/libstdc++-"..., 58) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "67", 2)                       = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "_ZN9__gnu_cxx27__verbose_termina"..., 44) = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "7966661", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "../../../../gcc-4.9.3/libstdc++-"..., 60) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "47", 2)                       = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "_ZN10__cxxabiv111__terminateEPFv"..., 34) = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "7966736", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "../../../../gcc-4.9.3/libstdc++-"..., 60) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "57", 2)                       = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "_ZSt9terminatev", 15)         = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "8292256", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "../../../../../gcc-4.9.3/libstdc"..., 57) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "92", 2)                       = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "execute_native_thread_routine", 29) = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547554950596", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547554950596", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "start_thread", 12)            = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547540290716", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547540290716", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "clone", 5)                    = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "18446744073709551615", 20)    = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "18446744073709551615", 20)    = -1 EPIPE (Broken pipe)
write(2, "??:0 ??", 7)                  = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "---END STACK TRACE---\n", 22) = -1 EPIPE (Broken pipe)
write(2, "---END ERROR---\n", 16)       = -1 EPIPE (Broken pipe)
write(2, "----\n", 5)                   = -1 EPIPE (Broken pipe)
rt_sigreturn()                          = -1 EPIPE (Broken pipe)
--- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=6927, si_uid=498} ---
write(2, "\n++++\n", 6)                 = -1 EPIPE (Broken pipe)
write(2, "[ERROR]\n", 8)                = -1 EPIPE (Broken pipe)
write(2, "---BEGIN ERROR----\n", 19)    = -1 EPIPE (Broken pipe)
write(2, "Signal: ", 8)                 = -1 EPIPE (Broken pipe)
write(2, "SIGPIPE", 7)                  = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "Description: ", 13)           = -1 EPIPE (Broken pipe)
write(2, "Broken pipe", 11)             = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "---BEGIN STACK TRACE---\n", 24) = -1 EPIPE (Broken pipe)
write(2, "4656540", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "/home/pfifer/producer-gh/amazon-"..., 88) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "104", 3)                      = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "_ZN3aws5utils9backtrace22stack_t"..., 50) = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "4655134", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "/home/pfifer/producer-gh/amazon-"..., 79) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "84", 2)                       = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "signal_handler", 14)          = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547554980095", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547554980095", 15)         = -1 EPIPE (Broken pipe)
write(2, "??:0 ??", 7)                  = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547540231373", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547540231373", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "write", 5)                    = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547539764114", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547539764114", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "_IO_new_file_write", 18)      = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547539766062", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547539766062", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "__GI__IO_file_xsputn", 20)    = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547539723292", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547539723292", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "fwrite", 6)                   = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "8343637", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "../../../../gcc-4.9.3/libstdc++-"..., 58) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "67", 2)                       = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "_ZN9__gnu_cxx27__verbose_termina"..., 44) = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "7966661", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "../../../../gcc-4.9.3/libstdc++-"..., 60) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "47", 2)                       = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "_ZN10__cxxabiv111__terminateEPFv"..., 34) = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "7966736", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "../../../../gcc-4.9.3/libstdc++-"..., 60) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "57", 2)                       = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "_ZSt9terminatev", 15)         = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "8292256", 7)                  = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "../../../../../gcc-4.9.3/libstdc"..., 57) = -1 EPIPE (Broken pipe)
write(2, ":", 1)                        = -1 EPIPE (Broken pipe)
write(2, "92", 2)                       = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "execute_native_thread_routine", 29) = -1 EPIPE (Broken pipe)
write(2, "(...)", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547554950596", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547554950596", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "start_thread", 12)            = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "140547540290716", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "140547540290716", 15)         = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "clone", 5)                    = -1 EPIPE (Broken pipe)
write(2, ":0 ??", 5)                    = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "18446744073709551615", 20)    = -1 EPIPE (Broken pipe)
write(2, " ", 1)                        = -1 EPIPE (Broken pipe)
write(2, "18446744073709551615", 20)    = -1 EPIPE (Broken pipe)
write(2, "??:0 ??", 7)                  = -1 EPIPE (Broken pipe)
write(2, "\n", 1)                       = -1 EPIPE (Broken pipe)
write(2, "---END STACK TRACE---\n", 22) = -1 EPIPE (Broken pipe)
write(2, "---END ERROR---\n", 16)       = -1 EPIPE (Broken pipe)
write(2, "----\n", 5)                   = -1 EPIPE (Broken pipe)
rt_sigreturn()                          = -1 EPIPE (Broken pipe)
--- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=6927, si_uid=498} ---
```

**Comment by pfifer** [comment]
The fix for this is in #77.  Killing the parent process of the kinesis_producer is always going to be bad.  For the future I'll look into adding a shutdown hook to the Java side which should attempt to reap the child before the Java process exits.

**Comment by pfifer** [comment]
Release 0.12.2 should fix this.  It's been published on Central, but still take some time to propagate.  Let me know if you still have this issue.
